File: packages/application/src/editorial/excerpt.ts
/**
 * The standfirst a listing shows under a headline.
 *
 * Articles carry no separate excerpt field, and adding one would mean asking
 * editors to write the same sentence twice. Taking the opening of the approved
 * revision is what a newsroom would do by hand, and it can never drift from
 * the published text because it IS the published text.
 *
 * Cuts on a word boundary so a listing never ends mid-word.
 */
export function excerptFrom(body: string, limit: number): string {
  const flat = body.replace(/\s+/gu, ' ').trim()
  if (flat.length <= limit) return flat

  const cut = flat.slice(0, limit)
  const lastSpace = cut.lastIndexOf(' ')

  // A body with no spaces inside the limit (a URL, say) has no word boundary
  // to respect; a hard cut beats returning the whole thing.
  return `${lastSpace > 0 ? cut.slice(0, lastSpace) : cut}\u2026`
}

File: services/api/internal/app/editorial/excerpt.go
package editorial

import "strings"

const wordsPerMinute = 200

// ExcerptFrom is the standfirst a listing shows under a headline.
//
// Articles carry no separate excerpt field. Taking the opening of the body is
// what a newsroom would do by hand. Cuts on a word boundary so a card never
// ends mid-word.
func ExcerptFrom(body string, limit int) string {
	flat := strings.Join(strings.Fields(body), " ")
	runes := []rune(flat)
	if len(runes) <= limit {
		return flat
	}

	cut := runes[:limit]
	lastSpace := -1
	for i, r := range cut {
		if r == ' ' {
			lastSpace = i
		}
	}
	if lastSpace > 0 {
		cut = cut[:lastSpace]
	}

	return string(cut) + "\u2026"
}

// ReadingTimeMinutes estimates the approved copy at a newsroom-standard pace.
func ReadingTimeMinutes(body string) int {
	words := len(strings.Fields(body))
	minutes := (words + wordsPerMinute - 1) / wordsPerMinute
	if minutes < 1 {
		return 1
	}

	return minutes
}
